cdap-data-fabric/src/test/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetServiceAuthorizationTest.java

```diff
@@ -0,0 +1,215 @@
+    // since Alice now is ADMIN for dsId1, she should be able to delete it
+    deleteAndAssertPrivileges(dsId1);
+    Assert.assertTrue(dsFramework.getInstances(NamespaceId.DEFAULT.toId()).isEmpty());
+    // Bob should still see 1 instance
+    SecurityRequestContext.setUserId(BOB.getName());
+    Assert.assertEquals(ImmutableSet.of(dsId2),
+                        summaryToDatasetIdSet(dsFramework.getInstances(NamespaceId.DEFAULT.toId())));
+    deleteAndAssertPrivileges(dsId2);
+  }
+
+  private Set<DatasetId> summaryToDatasetIdSet(Collection<DatasetSpecificationSummary> datasetSpecs) {
+    Collection<DatasetId> datasetIds =
+      Collections2.transform(datasetSpecs, new Function<DatasetSpecificationSummary, DatasetId>() {
+        @Override
+        public DatasetId apply(DatasetSpecificationSummary input) {
+          return NamespaceId.DEFAULT.dataset(input.getName());
+        }
+      });
+    return ImmutableSet.copyOf(datasetIds);
+  }
+
+  private void addInstanceAndAssertPrivileges(Principal principal, DatasetId dsId) throws Exception {
+    Set<Privilege> before = authorizer.listPrivileges(principal);
+    dsFramework.addInstance(Table.class.getName(), dsId.toId(), DatasetProperties.EMPTY);
+    Set<Privilege> after = authorizer.listPrivileges(principal);
+    Assert.assertTrue(after.containsAll(before));
+    Assert.assertEquals(ImmutableSet.of(new Privilege(dsId, Action.ALL)),
+                        Sets.difference(after, before).immutableCopy());
+  }
+
+  private void deleteAndAssertPrivileges(final DatasetId dsId) throws Exception {
+    dsFramework.deleteInstance(dsId.toId());
+    Predicate<Privilege> dsIdFilter = new Predicate<Privilege>() {
+      @Override
+      public boolean apply(Privilege input) {
+        return input.getEntity().equals(dsId);
+      }
+    };
+    Assert.assertTrue(Sets.filter(authorizer.listPrivileges(ALICE), dsIdFilter).isEmpty());
+    Assert.assertTrue(Sets.filter(authorizer.listPrivileges(BOB), dsIdFilter).isEmpty());
+  }
+
+  private void grantAndAssertSuccess(EntityId entityId, Principal principal, Set<Action> actions) throws Exception {
+    Set<Privilege> existingPrivileges = authorizer.listPrivileges(principal);
+    authorizer.grant(entityId, principal, actions);
+    ImmutableSet.Builder<Privilege> expectedPrivilegesAfterGrant = ImmutableSet.builder();
+    for (Action action : actions) {
+      expectedPrivilegesAfterGrant.add(new Privilege(entityId, action));
+    }
+    Assert.assertEquals(Sets.union(existingPrivileges, expectedPrivilegesAfterGrant.build()),
+                        authorizer.listPrivileges(principal));
+  }
+
+  private void assertAuthorizationFailure(UnauthorizedOperationExecutor operation, String failureMsg) throws Exception {
+    try {
+      operation.execute();
+      Assert.fail(failureMsg);
+    } catch (UnauthorizedException expected) {
+      // expected
+    } catch (DatasetManagementException e) {
```

Comment:
For authorization failures, when do we get UnauthorizedException thrown vs, DatasetManagementException thrown?